In [ ]:
import glob
import time
import pandas as pd
# from xml.dom import minidom
from nltk import ngrams
from nltk.tokenize import sent_tokenize
import nltk
nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')
from nltk.stem import PorterStemmer
from nltk.stem import PorterStemmer
from nltk.tokenize import sent_tokenize, word_tokenize
from nltk.stem import WordNetLemmatizer
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...


In [ ]:
import keras

In [ ]:
import nltk
import pandas as pd
import os

In [ ]:
# Xử lí các dữ liệu cho cùng một format

def clean_sqli_data(data):

    for i in range(len(data)):

        data[i]=data[i].replace('\n', '')
        data[i]=data[i].replace('%20', ' ')
        data[i]=data[i].replace('=', ' = ')
        data[i]=data[i].replace('((', ' (( ')
        data[i]=data[i].replace('))', ' )) ')
        data[i]=data[i].replace('(', ' ( ')
        data[i]=data[i].replace(')', ' ) ')

    return data

In [ ]:
path='/content/sample_data/sqli.txt'

In [ ]:
# đọc dữ liệu từ file

sql_lines_fuzzing=[]
f = open(path, "r")
for x in f:
    sql_lines_fuzzing.append(x)

In [ ]:
sql_lines_fuzzing=clean_sqli_data(sql_lines_fuzzing)

In [ ]:
sql_lines_fuzzing[:15]

["'sqlvuln",
 "'+sqlvuln",
 'sqlvuln;',
 ' ( sqlvuln ) ',
 "a' or 1 = 1--",
 '"a"" or 1 = 1--"',
 ' or a  =  a',
 "a' or 'a'  =  'a",
 '1 or 1 = 1',
 "a' waitfor delay '0:0:10'--",
 "1 waitfor delay '0:0:10'--",
 'declare @q nvarchar  ( 4000 )  select @q  = ',
 '0x770061006900740066006F0072002000640065006C00610079002000270030003A0030003A',
 '0',
 '031003000270000']

In [ ]:
path='/content/sample_data/camoufl4g3.txt'

In [ ]:
# Đọc dữ liệu

sql_lines_camoufl4g3=[]
f = open(path, "r")
for x in f:
    sql_lines_camoufl4g3.append(x)

In [ ]:
sql_lines_camoufl4g3[:15] # Dữ liệu trước khi làm sạch

["'-'\n",
 "' '\n",
 "'&'\n",
 "'^'\n",
 "'*'\n",
 "' or ''-'\n",
 "' or '' '\n",
 "' or ''&'\n",
 "' or ''^'\n",
 "' or ''*'\n",
 '"-"\n',
 '" "\n',
 '"&"\n',
 '"^"\n',
 '"*"\n']

In [ ]:
sql_lines_camoufl4g3=clean_sqli_data(sql_lines_camoufl4g3)

In [ ]:
sql_lines_camoufl4g3[:15]  # dữ liệu sau khi làm sạch

["'-'",
 "' '",
 "'&'",
 "'^'",
 "'*'",
 "' or ''-'",
 "' or '' '",
 "' or ''&'",
 "' or ''^'",
 "' or ''*'",
 '"-"',
 '" "',
 '"&"',
 '"^"',
 '"*"']

In [ ]:
path='/content/sample_data/libinjection-bypasses.txt'

In [ ]:
sql_lines_bypasses=[]
f = open(path, "r")
for x in f:
    sql_lines_bypasses.append(x)

In [ ]:
sql_lines_bypasses[:5]

['&()o1: select * from users where id=1 or (\\)=1 union select 1,@@VERSION -- 1\n',
 '&(.)o: select * from users where id=1 or (\\.)=1 union select 1,@@VERSION -- 1\n',
 '&(1&1: select * from users where id=1 or (\\+)=1 or 1=1 -- 1\n',
 '&(1)o: select * from users where id=1 or (1)=1 union select 1,banner from v$version where rownum=1 -- 1\n',
 '&(1UE: select * from users where id=1 or (\\+)=1 union select 1,@@VERSION -- 1\n']

In [ ]:
print(len(sql_lines_bypasses))

474


In [ ]:
sql_lines_bypasses=clean_sqli_data(sql_lines_bypasses)

In [ ]:
sql_lines_bypasses[:5]

['& (  ) o1: select * from users where id = 1 or  ( \\ )  = 1 union select 1,@@VERSION -- 1',
 '& ( . ) o: select * from users where id = 1 or  ( \\. )  = 1 union select 1,@@VERSION -- 1',
 '& ( 1&1: select * from users where id = 1 or  ( \\+ )  = 1 or 1 = 1 -- 1',
 '& ( 1 ) o: select * from users where id = 1 or  ( 1 )  = 1 union select 1,banner from v$version where rownum = 1 -- 1',
 '& ( 1UE: select * from users where id = 1 or  ( \\+ )  = 1 union select 1,@@VERSION -- 1']

In [ ]:
# Loại bỏ phần tiền tố bắt đầu bằng dấu &(*)* (hoặc các ký tự trước dấu :)

for i in range(len(sql_lines_bypasses)):
    sentence=sql_lines_bypasses[i]
    sql_lines_bypasses[i]=sentence.split(':')[1]


In [ ]:
sql_lines_bypasses[:5]

[' select * from users where id = 1 or  ( \\ )  = 1 union select 1,@@VERSION -- 1',
 ' select * from users where id = 1 or  ( \\. )  = 1 union select 1,@@VERSION -- 1',
 ' select * from users where id = 1 or  ( \\+ )  = 1 or 1 = 1 -- 1',
 ' select * from users where id = 1 or  ( 1 )  = 1 union select 1,banner from v$version where rownum = 1 -- 1',
 ' select * from users where id = 1 or  ( \\+ )  = 1 union select 1,@@VERSION -- 1']

In [ ]:
path='/content/sample_data/sqli_owasp_2.txt'

In [ ]:
# read data from file

sql_lines_owasp=[]
f = open(path, "r")
for x in f:
    sql_lines_owasp.append(x)

In [ ]:
sql_lines_owasp[0:15] # trước khi được làm sạch

['a\n',
 "a' \n",
 "a' --\n",
 "a' or 1=1; --\n",
 '@\n',
 '?\n',
 "' and 1=0) union all\n",
 '? or 1=1 --\n',
 "x' and userid is NULL; --\n",
 "x' and email is NULL; --\n",
 "anything' or 'x'='x\n",
 "x' and 1=(select count(*) from tabname); --\n",
 "x' and members.email is NULL; --\n",
 "x' or full_name like '%bob%\n",
 '23 or 1=1; --\n']

In [ ]:
print(len(sql_lines_owasp))

167


In [ ]:
sql_lines_owasp=clean_sqli_data(sql_lines_owasp)

In [ ]:
sql_lines_owasp[:15] # sau khi được làm sạch

['a',
 "a' ",
 "a' --",
 "a' or 1 = 1; --",
 '@',
 '?',
 "' and 1 = 0 )  union all",
 '? or 1 = 1 --',
 "x' and userid is NULL; --",
 "x' and email is NULL; --",
 "anything' or 'x' = 'x",
 "x' and 1 =  ( select count ( * )  from tabname ) ; --",
 "x' and members.email is NULL; --",
 "x' or full_name like '%bob%",
 '23 or 1 = 1; --']

In [ ]:
path='/content/sample_data/Generic-SQLi.txt'

In [ ]:
sql_lines_Generic=[]
f = open(path, "r")
for x in f:
    sql_lines_Generic.append(x)

In [ ]:
sql_lines_Generic[:15]

[")%20or%20('x'='x\n",
 '%20or%201=1\n',
 "; execute immediate 'sel' || 'ect us' || 'er'\n",
 'benchmark(10000000,MD5(1))#\n',
 'update\n',
 '";waitfor delay \'0:0:__TIME__\'--\n',
 '1) or pg_sleep(__TIME__)--\n',
 '||(elt(-3+5,bin(15),ord(10),hex(char(45))))\n',
 '"hi"") or (""a""=""a"\n',
 'delete\n',
 'like\n',
 '" or sleep(__TIME__)#\n',
 'pg_sleep(__TIME__)--\n',
 '*(|(objectclass=*))\n',
 'declare @q nvarchar (200) 0x730065006c00650063 ...\n']

In [ ]:
print(len(sql_lines_Generic))

268


In [ ]:
sql_lines_Generic=clean_sqli_data(sql_lines_Generic)

In [ ]:
sql_lines_Generic[:15]

[" )  or  ( 'x' = 'x",
 ' or 1 = 1',
 "; execute immediate 'sel' || 'ect us' || 'er'",
 'benchmark ( 10000000,MD5 ( 1  )  )  #',
 'update',
 '";waitfor delay \'0:0:__TIME__\'--',
 '1 )  or pg_sleep ( __TIME__ ) --',
 '|| ( elt ( -3+5,bin ( 15 ) ,ord ( 10 ) ,hex ( char ( 45  )  )    )  )  ',
 '"hi"" )  or  ( ""a"" = ""a"',
 'delete',
 'like',
 '" or sleep ( __TIME__ ) #',
 'pg_sleep ( __TIME__ ) --',
 '* ( | ( objectclass = *  )  )  ',
 'declare @q nvarchar  ( 200 )  0x730065006c00650063 ...']

In [ ]:

# loại bỏ các từ dừng (stopwords) khỏi văn bản đầu vào, từ dừng là những từ như "the", "and", "is", "in", "of"

stop_words = set(stopwords.words('english'))

def fun_remove_stop_words(posts):

    filtered=''

    for x in posts.split(' '):
        if x not in stop_words:
            filtered+=' '+x

    return filtered

In [ ]:
path='/content/sample_data/'
file="plain.txt"

In [ ]:
df = pd.read_csv(os.path.join(path,file), sep='Aw3s0meSc0t7', names=['benign'], header=None, engine='python')

In [ ]:
df.head()

benign
0                           Add plain text here
1                    “Ne te quaesiveris extra.”
2   “Man is his own star; and the soul that can
3           Render an honest and a perfect man,
4  Commands all light, all influence, all fate;

In [ ]:
plain_text=df['benign'].values

In [ ]:
plain_text[:5]

array(['Add plain text here', '“Ne te quaesiveris extra.”',
       '“Man is his own star; and the soul that can',
       'Render an honest and a perfect man,',
       'Commands all light, all influence, all fate;'], dtype=object)

In [ ]:
plain_text=plain_text[:-22] #  loại bỏ 22 ký tự là URLs

In [ ]:
len(plain_text)

3672

In [ ]:
# chuyển đổi sang dạng string
data=''
for x in plain_text:
    data+=" " + x

In [ ]:
type(data)

str

In [ ]:
data=fun_remove_stop_words(data)  # loại bỏ từ dừng
data=data.split('.')              # Chia chuỗi thành các câu
# "This is a test. How are you?." = ["This is a test", " How are you?"]

In [ ]:
# phân tách html có khoảng trống

for i in range(len(data)):
    data[i]=data[i].replace('<', ' <')
    data[i]=data[i].replace('>', '> ')
    data[i]=data[i].replace('=', ' = ')

In [ ]:
data[:5] # dữ liệu sau khi chỉnh sửa

['  Add plain text “Ne te quaesiveris extra',
 '” “Man star; soul Render honest perfect man, Commands light, influence, fate; Nothing falls early late',
 ' Our acts angels are, good ill, Our fatal shadows walk us still',
 '” Epilogue Beaumont Fletcher’s Honest Man’s Fortune Cast bantling rocks, Suckle she-wolf’s teat; Wintered hawk fox, Power speed hands feet',
 ' I read day verses written eminent painter original conventional']

In [ ]:
print("Tổng: %2i" %len(data))

Tổng: 3072


In [ ]:
path='/content/sample_data/benign_for_training.txt'
benign_data=[]
f = open(path, "r")
for x in f:
    benign_data.append(x)


In [ ]:
path='/content/sample_data/sqli_for_training.txt'
sqli_data=[]
f = open(path, "r")
for x in f:
    sqli_data.append(x)


In [ ]:
len(benign_data)

440

In [ ]:
benign_sentence=[]
for i in benign_data:
    sentences=i.split('.')

    for sentence in sentences:
        benign_sentence.append(sentence)

In [ ]:
len(benign_sentence)

1115

In [ ]:
print("Tổng : " ,len(benign_sentence)+len(data))

Tổng :  4187


In [ ]:
len(sqli_data)

290

In [ ]:
print(f"Số lượng các SQL fuzzing : {len(sql_lines_fuzzing)}  camoufl4gs : {len(sql_lines_camoufl4g3)} parsed : {len(sql_lines_bypasses)} owasp : {len(sql_lines_owasp)} generic : {len(sql_lines_Generic)} \n total sql injection data : {len(sql_lines_Generic)+len(sql_lines_owasp)+len(sql_lines_bypasses)+len(sql_lines_camoufl4g3)+len(sql_lines_fuzzing)} ")

Số lượng các SQL fuzzing : 142  camoufl4gs : 77 parsed : 474 owasp : 167 generic : 268 
 total sql injection data : 1128 


In [ ]:
# Gộp lại 1 biến
all_sqli_sentence=sql_lines_owasp+sql_lines_bypasses+sql_lines_camoufl4g3+sql_lines_fuzzing+sql_lines_Generic

Thay thế các giá trị số (numeric values) trong một danh sách các câu SQL bằng một từ khóa "numeric".

In [ ]:
len(all_sqli_sentence)

1128

In [ ]:
# thay thế giá trị numeric thành keyword 'numeric'
def optional_numeric_to_numeric(all_sqli_sentence):

    for i in range(len(all_sqli_sentence)):


        all_sqli_sentence[i]=all_sqli_sentence[i].replace('1 ', 'numeric')
        all_sqli_sentence[i]=all_sqli_sentence[i].replace(' 1', 'numeric')
        all_sqli_sentence[i]=all_sqli_sentence[i].replace("'1 ", "'numeric ")
        all_sqli_sentence[i]=all_sqli_sentence[i].replace(" 1'", " numeric'")
        all_sqli_sentence[i]=all_sqli_sentence[i].replace('1,', 'numeric,')
        all_sqli_sentence[i]=all_sqli_sentence[i].replace("1\ ", 'numeric')
        all_sqli_sentence[i]=all_sqli_sentence[i].replace("‘1", '‘numeric')
        all_sqli_sentence[i]=all_sqli_sentence[i].replace(" 2 ", " numeric ")
        all_sqli_sentence[i]=all_sqli_sentence[i].replace(' 3 ', ' numeric ')
        all_sqli_sentence[i]=all_sqli_sentence[i].replace(' 3--', ' numeric--')
        all_sqli_sentence[i]=all_sqli_sentence[i].replace(" 4 ", ' numeric ')
        all_sqli_sentence[i]=all_sqli_sentence[i].replace(" 5 ", ' numeric ')
        all_sqli_sentence[i]=all_sqli_sentence[i].replace(' 6 ', ' numeric ')
        all_sqli_sentence[i]=all_sqli_sentence[i].replace(" 7 ", ' numeric ')
        all_sqli_sentence[i]=all_sqli_sentence[i].replace(" 8 ", ' numeric ')
        all_sqli_sentence[i]=all_sqli_sentence[i].replace('1234', ' numeric ')
        all_sqli_sentence[i]=all_sqli_sentence[i].replace("22", ' numeric ')
        all_sqli_sentence[i]=all_sqli_sentence[i].replace(" 8 ", ' numeric ')
        all_sqli_sentence[i]=all_sqli_sentence[i].replace(" 200 ", ' numeric ')
        all_sqli_sentence[i]=all_sqli_sentence[i].replace("23 ", ' numeric ')
        all_sqli_sentence[i]=all_sqli_sentence[i].replace('"1', '"numeric')
        all_sqli_sentence[i]=all_sqli_sentence[i].replace('1"', '"numeric')
        all_sqli_sentence[i]=all_sqli_sentence[i].replace("7659", 'numeric')
        all_sqli_sentence[i]=all_sqli_sentence[i].replace(" 37 ", ' numeric ')
        all_sqli_sentence[i]=all_sqli_sentence[i].replace(" 45 ", ' numeric ')

    return all_sqli_sentence


In [ ]:
all_sqli_sentence

['a',
 "a' ",
 "a' --",
 "a' or 1 = 1; --",
 '@',
 '?',
 "' and 1 = 0 )  union all",
 '? or 1 = 1 --',
 "x' and userid is NULL; --",
 "x' and email is NULL; --",
 "anything' or 'x' = 'x",
 "x' and 1 =  ( select count ( * )  from tabname ) ; --",
 "x' and members.email is NULL; --",
 "x' or full_name like '%bob%",
 '23 or 1 = 1; --',
 "'; exec master..xp_cmdshell 'ping 172.10.1.255'--",
 'a',
 '1 or 1 = 1',
 "1' or '1' = '1",
 "1 and user_name (  )   =  'dbo'",
 '1',
 "1'1",
 '1 exec sp_  ( or exec xp_ ) ',
 '1 and 1 = 1',
 "1' and 1 =  ( select count ( * )  from tablenames ) ; --",
 '1',
 "1 and user_name (  )   =  'dbo'",
 "\\'; desc users; --",
 "1\\'1",
 "1' and non_existant_table  =  '1",
 "' or username is not NULL or username  =  '",
 "1 and ascii ( lower ( substring  (  (  select top 1 name from sysobjects where xtype = 'u' ) , 1, 1  )  )   )  > 116",
 "1 union all select 1,2,3,4,5,6,name from sysobjects where xtype  =  'u' --",
 '1 uni/**/on select all from where',
 '’ or ‘1’ =

In [ ]:
import pandas as pd

In [ ]:
# gán label cho sqli

values=[]
for i in all_sqli_sentence:
    values.append((i,1))

In [ ]:
# gán label cho begin data

for i in data:
    values.append((i,0))

In [ ]:
len(all_sqli_sentence)+len(data)

4200

In [ ]:
len(values)

4200

In [ ]:
for i in benign_sentence:
    values.append((i,0))

In [ ]:
len(values)

5315

In [ ]:
for i in sqli_data:
    values.append((i,1))

In [ ]:
len(values)

5605

In [ ]:
values[1]

("a' ", 1)

In [ ]:
# chuyển sang dạng dataframe

df=pd.DataFrame(values,columns=['Sentence','Label'])



In [ ]:
df.head() # Kiểm tra lại dữ liệu

Sentence  Label
0                a      1
1              a'       1
2            a' --      1
3  a' or 1 = 1; --      1
4                @      1

In [ ]:
# Sau khi chuyển đổi, lưu lại thành file csv
df.to_csv('sqli.csv', index=False, encoding='utf-16')

In [ ]:
df=pd.read_csv('/content/sample_data/sqli.csv',encoding='utf-16')

**min_df=2**: Từ chỉ được đưa vào nếu nó xuất hiện ít nhất 2 lần trong tập dữ liệu. Điều này giúp loại bỏ các từ ít gặp, có thể không mang nhiều thông tin.

**max_df=0.7**: Từ chỉ được đưa vào nếu nó xuất hiện ở không quá 70% số tài liệu (câu). Điều này giúp loại bỏ các từ quá phổ biến như "the", "is", v.v.

**max_features=4096**: Số lượng từ tối đa mà bộ vector hóa sẽ giữ lại. Trong trường hợp này, chỉ giữ lại 4096 từ phổ biến nhất.

**stop_words=stopwords.words('english')**: Loại bỏ các từ dừng (stopwords) trong tiếng Anh như "the", "is", "at", v.v., vì chúng thường không mang nhiều giá trị phân biệt trong các tác vụ phân tích văn bản.

**df['Sentence']**: Truy cập cột Sentence trong DataFrame df, nơi chứa các câu văn bản cần vector hóa.

**.values.astype('U')**: Chuyển đổi các câu văn bản thành kiểu dữ liệu string (ký tự).

**fit_transform()**: Áp dụng quá trình huấn luyện và chuyển đổi dữ liệu từ dạng văn bản thành ma trận các đặc trưng số. Phương thức này học từ các câu văn trong df['Sentence'] và trả về một ma trận số, trong đó mỗi cột đại diện cho tần suất của một từ trong văn bản.

**.toarray()**: Chuyển ma trận sparse (rỗng) thành một ma trận dày đặc (dense), giúp dễ dàng thao tác với dữ liệu ở dạng mảng numpy.

In [ ]:
# vector hóa dữ liệu

from sklearn.feature_extraction.text import CountVectorizer
vectorizer = CountVectorizer( min_df=2, max_df=0.7, max_features=4096, stop_words=stopwords.words('english'))
posts = vectorizer.fit_transform(df['Sentence'].values.astype('U')).toarray()



**posts** là một mảng 2 chiều chứa các vector hóa của các câu văn trong cột Sentence. Mỗi câu được chuyển thành một vector, trong đó mỗi phần tử trong vector đại diện cho số lần xuất hiện của một từ trong câu đó.

Số lượng từ (vocabulary size) được giới hạn bởi các tham số như max_features=4096, min_df=2, và max_df=0.7.

In [ ]:
posts.shape

(5605, 4096)

5605: Có 5605 mẫu dữ liệu (vẫn là số lượng câu như trước).

64 và 64: Mỗi mẫu dữ liệu được chuyển thành một ma trận 64x64, có thể là hình ảnh 2D hoặc một ma trận có kích thước 64x64.

1: Chỉ có 1 kênh màu (có thể là ảnh đơn sắc hoặc grayscale), điều này có nghĩa là mỗi ma trận 64x64 chỉ có một lớp dữ liệu.

In [ ]:
posts.shape=(5605, 64, 64, 1)

In [ ]:
posts.shape

(5605, 64, 64, 1)

In [ ]:
X=posts

In [ ]:
y=df['Label']

X: Dữ liệu đầu vào (features), trong trường hợp này là các vector đã được vector hóa của các câu SQL. Mỗi câu được biểu diễn dưới dạng một ma trận 2D với kích thước (5605, 64, 64, 1) (tức là 5605 mẫu và mỗi mẫu có kích thước 64x64 với 1 kênh màu).

y: Nhãn mục tiêu (target labels), là cột Label trong DataFrame df. Đây có thể là nhãn phân loại (ví dụ: "SQL injection" hoặc "benign") cho mỗi câu trong tập dữ liệu.

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
# phân chia training dữ liệu sang các mẫu X và Y

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

train_test_split: Đây là phương thức từ scikit-learn dùng để chia dữ liệu thành hai phần: một phần cho việc huấn luyện (train) và một phần cho việc kiểm tra (test).

X_train và y_train: Dữ liệu huấn luyện và nhãn huấn luyện (80% của tổng dữ liệu, theo test_size=0.2).

X_test và y_test: Dữ liệu kiểm tra và nhãn kiểm tra (20% của tổng dữ liệu).

test_size=0.2: Xác định tỷ lệ dữ liệu kiểm tra (20% của tổng số dữ liệu).

random_state=42: Đảm bảo rằng việc chia dữ liệu được tái lập (reproducible) mỗi lần chạy mã (giúp tạo ra cùng một tập dữ liệu huấn luyện và kiểm tra nếu chạy lại với cùng một giá trị random_state).

In [ ]:
trainX=X_train.copy()
trainX.shape=(X_train.shape[0],trainX.shape[1]*trainX.shape[2])
testX=X_test.copy()
testX.shape=(testX.shape[0],testX.shape[1]*testX.shape[2])


trainX = X_train.copy() và testX = X_test.copy(): Tạo các bản sao của X_train và X_test để tránh việc thay đổi trực tiếp dữ liệu gốc.

trainX.shape = (X_train.shape[0], trainX.shape[1] * trainX.shape[2]):

Dữ liệu huấn luyện trainX có hình dạng (5605, 64, 64, 1), nhưng để phù hợp với các mô hình học máy hoặc các thuật toán yêu cầu dữ liệu đầu vào là một mảng 2 chiều (chẳng hạn như các mô hình fully connected neural networks), bạn cần thay đổi hình dạng này.

trainX.shape[0] giữ nguyên số lượng mẫu (5605), và trainX.shape[1] * trainX.shape[2] nhân kích thước của mỗi mẫu (64x64) thành một chiều đơn, làm cho mỗi mẫu trở thành một vector với 4096 đặc trưng.

Sau khi thay đổi, trainX sẽ có hình dạng (5605, 4096), tức là mỗi câu trong X_train sẽ trở thành một vector 4096 chiều.

testX.shape = (testX.shape[0], testX.shape[1] * testX.shape[2]): Tương tự như trên, dữ liệu kiểm tra X_test cũng được thay đổi để có dạng 2D (sau khi nhân các chiều của mỗi mẫu).

 # Thực hiện các bước huấn luyện và dự đoán sử dụng hai mô hình học máy: Naive Bayes và Support Vector Machine (SVM).

In [ ]:
from sklearn.naive_bayes import GaussianNB

gnb = GaussianNB()

gnb.fit(trainX, y_train)

pred_gnb = gnb.predict(testX)

GaussianNB là mô hình Naive Bayes sử dụng phân phối chuẩn (Gaussian distribution). Đây là một mô hình phân loại dựa trên định lý Bayes, giả định rằng các đặc trưng của dữ liệu (features) có phân phối chuẩn (Gaussian).

*   gnb.fit(trainX, y_train): huấn luyện mô hình Naive Bayes trên tập huấn luyện trainX (dữ liệu đầu vào) và y_train (nhãn mục tiêu).
*   pred_gnb = gnb.predict(testX): Sau khi mô hình được huấn luyện, predict() được gọi để dự đoán nhãn cho tập kiểm tra testX.


In [ ]:
# SVM
from sklearn.svm import SVC
clf = SVC(gamma='auto')
clf.fit(trainX, y_train)
pred_svm=clf.predict(testX)

SVC là một trong những thuật toán của Support Vector Machine (SVM), được sử dụng để phân loại dữ liệu. SVM tìm kiếm một siêu phẳng (hyperplane) tối ưu để phân tách các lớp dữ liệu.

gamma='auto': Tham số này xác định giá trị gamma trong thuật toán SVM. Giá trị auto có nghĩa là gamma sẽ được tính từ số lượng đặc trưng trong dữ liệu.

clf.fit(trainX, y_train): Dòng này huấn luyện mô hình SVM trên tập huấn luyện trainX (dữ liệu đầu vào) và y_train (nhãn mục tiêu).

pred_svm = clf.predict(testX): Sau khi mô hình được huấn luyện, predict() được gọi để dự đoán nhãn cho tập kiểm tra testX.

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

neigh = KNeighborsClassifier(n_neighbors=3)
neigh.fit(trainX, y_train)

pred_knn = neigh.predict(testX)

KNeighborsClassifier: Là thuật toán K-Nearest Neighbors (KNN), một mô hình phân loại không tham số. KNN phân loại một mẫu dữ liệu bằng cách tìm K điểm dữ liệu gần nhất (neighbors) với mẫu cần phân loại và lấy nhãn của phần lớn các điểm gần nhất đó.

n_neighbors=3: Số lượng điểm gần nhất (K=3) được xem xét để phân loại một mẫu dữ liệu.

neigh.fit(trainX, y_train): Dòng này huấn luyện mô hình KNN trên tập huấn luyện trainX và nhãn y_train.

pred_knn = neigh.predict(testX): Dòng này dự đoán nhãn cho tập kiểm tra testX.

In [ ]:
from sklearn import tree

dt = tree.DecisionTreeClassifier()
dt = dt.fit(trainX, y_train)

pred_dt = dt.predict(testX)

DecisionTreeClassifier: Là mô hình Cây quyết định (Decision Tree), một thuật toán phân loại sử dụng cây quyết định để phân loại dữ liệu. Cây quyết định xây dựng một mô hình phân loại thông qua các quyết định phân nhánh dựa trên các đặc trưng.

dt.fit(trainX, y_train): Dòng này huấn luyện mô hình cây quyết định trên tập huấn luyện trainX và nhãn y_train.

pred_dt = dt.predict(testX): Dòng này dự đoán nhãn cho tập kiểm tra testX.

In [ ]:
X_train.shape

(4484, 64, 64, 1)

Cài đặt thêm các thư viện còn thiếu

In [ ]:
!pip install scikit-learn


In [ ]:
!pip install tensorflow

In [ ]:
!pip install Keras-Preprocessing

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.6/42.6 kB 1.8 MB/s eta 0:00:00


In [ ]:
!pip install tensorflow-text

In [ ]:
!pip install scikeras

# CNN

In [ ]:
import tensorflow as tf
from keras.models import Sequential
from keras import layers
from tensorflow.keras.preprocessing.text import Tokenizer
from scikeras.wrappers import KerasRegressor

In [ ]:
model=tf.keras.models.Sequential([

    tf.keras.layers.Conv2D(64, (3,3), activation=tf.nn.relu, input_shape=(64,64,1)),
    tf.keras.layers.MaxPooling2D(2,2),

    tf.keras.layers.Conv2D(128,(3,3), activation=tf.nn.relu),
    tf.keras.layers.MaxPooling2D(2,2),

    tf.keras.layers.Conv2D(256,(3,3), activation='relu'),
    tf.keras.layers.MaxPooling2D(2,2),

    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(256, activation='relu'),
    tf.keras.layers.Dense(128,activation='relu'),
    tf.keras.layers.Dense(64, activation=tf.nn.relu),
    tf.keras.layers.Dense(1, activation='sigmoid')
])



/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
model.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                      │ (None, 62, 62, 64)          │             640 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d (MaxPooling2D)         │ (None, 31, 31, 64)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv2d_1 (Conv2D)                    │ (None, 29, 29, 128)         │          73,856 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_1 (MaxPooling2D)       │ (None, 14, 14, 128)         │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv2d_2 (Conv2D)                    │ (None, 12, 12, 256)         │         295,168 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_2 (MaxPooling2D)       │ (None, 6, 6, 256)           │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten (Flatten)                    │ (None, 9216)                │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense (Dense)                        │ (None, 256)                 │       2,359,552 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 128)                 │          32,896 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 64)                  │           8,256 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_3 (Dense)                      │ (None, 1)                   │              65 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 2,770,433 (10.57 MB)

 Trainable params: 2,770,433 (10.57 MB)

 Non-trainable params: 0 (0.00 B)

loss='binary_crossentropy': Chọn hàm mất mát (loss function) là binary crossentropy, phù hợp cho bài toán phân loại nhị phân. Hàm mất mát này đo lường sự khác biệt giữa nhãn dự đoán và nhãn thực tế, và mục tiêu là giảm giá trị hàm mất mát.

optimizer='adam': Chọn Adam làm thuật toán tối ưu hóa, đây là một trong những thuật toán tối ưu phổ biến và hiệu quả, đặc biệt khi làm việc với các mô hình học sâu.

metrics=['accuracy']: Đo lường hiệu suất của mô hình bằng độ chính xác (accuracy), tức là tỷ lệ dự đoán đúng trên tổng số mẫu.

In [ ]:
classifier_nn = model.fit(X_train,y_train,
                    epochs=10,
                    verbose=True,
                    validation_data=(X_test, y_test),
                    batch_size=128)


Epoch 1/10
36/36 ━━━━━━━━━━━━━━━━━━━━ 109s 3s/step - accuracy: 0.8096 - loss: 0.4135 - val_accuracy: 0.9179 - val_loss: 0.2159
Epoch 2/10
36/36 ━━━━━━━━━━━━━━━━━━━━ 116s 2s/step - accuracy: 0.9158 - loss: 0.1963 - val_accuracy: 0.9215 - val_loss: 0.1941
Epoch 3/10
36/36 ━━━━━━━━━━━━━━━━━━━━ 78s 2s/step - accuracy: 0.9283 - loss: 0.1508 - val_accuracy: 0.9269 - val_loss: 0.1864
Epoch 4/10
36/36 ━━━━━━━━━━━━━━━━━━━━ 86s 2s/step - accuracy: 0.9371 - loss: 0.1252 - val_accuracy: 0.9322 - val_loss: 0.1951
Epoch 5/10
36/36 ━━━━━━━━━━━━━━━━━━━━ 78s 2s/step - accuracy: 0.9416 - loss: 0.1120 - val_accuracy: 0.9358 - val_loss: 0.1932
Epoch 6/10
36/36 ━━━━━━━━━━━━━━━━━━━━ 81s 2s/step - accuracy: 0.9504 - loss: 0.0951 - val_accuracy: 0.9447 - val_loss: 0.1430
Epoch 7/10
36/36 ━━━━━━━━━━━━━━━━━━━━ 87s 2s/step - accuracy: 0.9489 - loss: 0.0942 - val_accuracy: 0.9367 - val_loss: 0.2417
Epoch 8/10
36/36 ━━━━━━━━━━━━━━━━━━━━ 89s 2s/step - accuracy: 0.9622 - loss: 0.0774 - val_accuracy: 0.9465 - val_los

In [ ]:
pred=model.predict(X_test)

36/36 ━━━━━━━━━━━━━━━━━━━━ 5s 125ms/step


In [ ]:
for i in range(len(pred)):
    if pred[i]>0.5:
        pred[i]=1
    elif pred[i]<=0.5:
        pred[i]=0

Chuyển đổi các giá trị dự doán về 1 hoặc 0. Tức là có hay không sql injection xảy ra.

In [ ]:
from sklearn.metrics import accuracy_score

In [ ]:
accuracy_score(y_test,pred)

0.9455842997323818

accuracy_score(y_test, pred):

y_test: Là nhãn thực tế của tập kiểm tra (test set), chứa các nhãn chính xác cho mỗi mẫu trong X_test.

pred: Là các nhãn dự đoán mà mô hình đã đưa ra, sau khi đã chuyển đổi các xác suất thành nhãn 0 hoặc 1 (như đã được thực hiện trong mã trước đó).

Hàm accuracy_score sẽ so sánh các giá trị trong y_test và pred, rồi tính toán tỷ lệ phần trăm các dự đoán đúng (tức là khi giá trị trong y_test bằng giá trị trong pred).

==> Mô hình dự đoán đúng khoảng 94.56% số mẫu trong tập kiểm tra.

In [ ]:
for i,j in zip(y_test,pred):
    print(i==j)

[ True]
[ True]
[ True]
[False]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[False]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[False]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[False]
[ True]
[ True]
[ True]
[ True]
[ True]
[False]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[False]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[ True]
[False]
[ True]
[ True]
[False]


So sánh kiểm tra giữa tập thực tế và tập dự đoán.

In [ ]:
from keras.models import load_model
import pickle

model.save('my_model_cnn.h5')
with open('vectorizer_cnn', 'wb') as fin:
    pickle.dump(vectorizer, fin)


Lưu lại model và vector

# Matrix Confusion

tp (True Positives): Số lượng mẫu mà mô hình dự đoán đúng là lớp "1" và thực tế là lớp "1".

tn (True Negatives): Số lượng mẫu mà mô hình dự đoán đúng là lớp "0" và thực tế là lớp "0".

fp (False Positives): Số lượng mẫu mà mô hình dự đoán sai là lớp "1", nhưng thực tế lại là lớp "0" (sai tích cực).

fn (False Negatives): Số lượng mẫu mà mô hình dự đoán sai là lớp "0", nhưng thực tế lại là lớp "1" (sai tiêu cực).

accuracy = (tp + tn) / (tp + tn + fp + fn):

Công thức tính độ chính xác (accuracy) là tỷ lệ giữa tổng số mẫu dự đoán đúng (cả lớp "1" và lớp "0") và tổng số mẫu tổng thể.

tp + tn: Số lượng mẫu dự đoán đúng (bao gồm cả đúng dương và đúng âm).

tp + tn + fp + fn: Tổng số mẫu (tất cả các dự đoán, đúng và sai).

Độ chính xác (accuracy) là tỷ lệ giữa số mẫu dự đoán đúng và tổng số mẫu, có thể hiểu là:

Accuracy = (Số mẫu dự đoán đúng) / (Tổng số mẫu)

![anh1](https://cdn.noron.vn/2018/07/08/93228024e0aa39e96ff82ad02ecfa224.jpg)

In [ ]:
def accuracy_function(tp,tn,fp,fn):

    accuracy = (tp+tn) / (tp+tn+fp+fn)

    return accuracy

In [ ]:
def precision_function(tp,fp):

    precision = tp / (tp+fp)

    return precision

In [ ]:
def recall_function(tp,fn):

    recall=tp / (tp+fn)

    return recall

In [ ]:
def confusion_matrix(truth,predicted):

    true_positive = 0
    true_negative = 0
    false_positive = 0
    false_negative = 0

    for true,pred in zip(truth,predicted):

        if true == 1:
            if pred == true:
                true_positive += 1
            elif pred != true:
                false_negative += 1

        elif true == 0:
            if pred == true:
                true_negative += 1
            elif pred != true:
                false_positive += 1

    accuracy=accuracy_function(true_positive, true_negative, false_positive, false_negative)
    precision=precision_function(true_positive, false_positive)
    recall=recall_function(true_positive, false_negative)

    return (accuracy,
            precision,
           recall)

In [ ]:
accuracy,precision,recall=confusion_matrix(y_test,pred)

In [ ]:
print("CNN \n Accuracy : {0} \n Precision : {1} \n Recall : {2}".format(accuracy, precision, recall))

CNN 
 Accuracy : 0.9393398751115076 
 Precision : 0.8549848942598187 
 Recall : 0.933993399339934


# Đánh giá các mô hình khác

In [ ]:
accuracy,precision,recall=confusion_matrix(y_test,pred_gnb)
print(" Mô hình Naive Bayes Accuracy : {0} \n Precision : {1} \n Recall : {2}".format(accuracy, precision, recall))

 For Naive Bayes Accuracy : 0.9562890276538805 
 Precision : 0.8825301204819277 
 Recall : 0.966996699669967


In [ ]:
accuracy,precision,recall=confusion_matrix(y_test,pred_svm)
print(" Mô hình SVM Accuracy : {0} \n Precision : {1} \n Recall : {2}".format(accuracy, precision, recall))

 For SVM Accuracy : 0.7975022301516503 
 Precision : 1.0 
 Recall : 0.2508250825082508


In [ ]:
accuracy,precision,recall=confusion_matrix(y_test,pred_knn)
print(" Mô hình KNN Accuracy : {0} \n Precision : {1} \n Recall : {2}".format(accuracy, precision, recall))

 For KNN Accuracy : 0.7154326494201606 
 Precision : 0.4831932773109244 
 Recall : 0.759075907590759


In [ ]:
accuracy,precision,recall=confusion_matrix(y_test,pred_dt)
print(" Mô hình Decision Tree Accuracy : {0} \n Precision : {1} \n Recall : {2}".format(accuracy, precision, recall))

 For Decision Tree Accuracy : 0.8661909009812667 
 Precision : 0.6750572082379863 
 Recall : 0.9735973597359736


In [1]:

import keras
from keras.models import load_model
import pickle


mymodel = tf.keras.models.load_model('/content/sample_data/my_model_additional_sqli_benign_data.h5')
myvectorizer = pickle.load(open("/content/sample_data/vectorizer_additional_sqli_benign_data", 'rb'))




def clean_data(input_val):

    input_val=input_val.replace('\n', '')
    input_val=input_val.replace('%20', ' ')
    input_val=input_val.replace('=', ' = ')
    input_val=input_val.replace('((', ' (( ')
    input_val=input_val.replace('))', ' )) ')
    input_val=input_val.replace('(', ' ( ')
    input_val=input_val.replace(')', ' ) ')
    input_val=input_val.replace('1 ', 'numeric')
    input_val=input_val.replace(' 1', 'numeric')
    input_val=input_val.replace("'1 ", "'numeric ")
    input_val=input_val.replace(" 1'", " numeric'")
    input_val=input_val.replace('1,', 'numeric,')
    input_val=input_val.replace(" 2 ", " numeric ")
    input_val=input_val.replace(' 3 ', ' numeric ')
    input_val=input_val.replace(' 3--', ' numeric--')
    input_val=input_val.replace(" 4 ", ' numeric ')
    input_val=input_val.replace(" 5 ", ' numeric ')
    input_val=input_val.replace(' 6 ', ' numeric ')
    input_val=input_val.replace(" 7 ", ' numeric ')
    input_val=input_val.replace(" 8 ", ' numeric ')
    input_val=input_val.replace('1234', ' numeric ')
    input_val=input_val.replace("22", ' numeric ')
    input_val=input_val.replace(" 8 ", ' numeric ')
    input_val=input_val.replace(" 200 ", ' numeric ')
    input_val=input_val.replace("23 ", ' numeric ')
    input_val=input_val.replace('"1', '"numeric')
    input_val=input_val.replace('1"', '"numeric')
    input_val=input_val.replace("7659", 'numeric')
    input_val=input_val.replace(" 37 ", ' numeric ')
    input_val=input_val.replace(" 45 ", ' numeric ')

    return input_val




def predict_sqli_attack():

    repeat=True

    beautify=''
    for i in range(20):
        beautify+= "="

    print(beautify)
    input_val=input("Dữ liệu kiểm tra : ")
    print(beautify)


    if input_val== '0':
        repeat=False

    input_val=clean_data(input_val)
    input_val=[input_val]

    input_val=myvectorizer.transform(input_val).toarray()

    input_val.shape=(1,64,64,1)

    result=mymodel.predict(input_val)

    # Print the divider with color
    print(beautify)

    if repeat == True:

        # In chữ đỏ nếu phát hiện SQL injection
        if result > 0.5:
            print("\033[91mCảnh báo !!! Phát hiện SQL injection\033[0m")  # Chữ đỏ

        # In chữ xanh nếu không phát hiện SQL injection
        elif result <= 0.5:
            print("Không ảnh hưởng")

        print(beautify)

        predict_sqli_attack()

    elif repeat == False:
        print( "Dừng hoạt động")


NameError: name 'tf' is not defined

In [ ]:
# check for sql injection cheat sheet





predict_sqli_attack()

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 40ms/step
Cảnh báo !!! Phát hiện SQL injection
